### PR TITLE
Making executor_traits N-nary

### LINKED ISSUE #2104: Make executor_traits N-nary

### PR DESCRIPTION
**Pull request by hkaiser**
This fixes #2104 

The breaking change is that the bulk- `async_execute` was renamed to `bulk_async_execute` and the bulk- `execute` was renamed to `bulk_execute`

### CODE REVIEW

File: hpx/parallel/executors/distribution_policy_executor.hpp [hunk 1]
```diff
@@ -60,53 +72,54 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
...
             !hpx::traits::is_action<F>::value
         >::type
-        apply_execute(F && f) const
+        apply_execute_impl(F && f, Ts && ... ts) const
         {
             typedef components::server::invoke_function_action<F> action_type;
-            policy_.template apply<action_type>(
-                threads::thread_priority_default, std::forward<F>(f));
+            policy_.template apply<action_type>(threads::thread_priority_default,
+                std::forward<F>(f), std::forward<Ts>(ts)...);
```
Comment (sithhell) [inline hpx/parallel/executors/distribution_policy_executor.hpp]: Do we really execute something that's not an action with a distribution policy?

File: hpx/parallel/executors/distribution_policy_executor.hpp [hunk 2]
```diff
@@ -60,53 +72,54 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
...
         // async_execute implementations
-        template <typename F>
+        template <typename F, typename ... Ts>
         typename std::enable_if<
-            !hpx::traits::is_action<F>::value,
+           !hpx::traits::is_action<F>::value,
             hpx::future<
                 typename detail::distribution_policy_execute_result<F>::type
             >
         >::type
```
Comment (sithhell) [inline hpx/parallel/executors/distribution_policy_executor.hpp]: One might think that this enable_if could be replaced with a static_assert in the function body. Maybe it gets clearer when I understand how to execute non-actions with a distribution policy

File: hpx/parallel/executors/distribution_policy_executor.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/parallel/executors/distribution_policy_executor.hpp]: Well, nothing has changed relatively to the implementation of distribution_policies. I'd be happy to discuss this on a separate PR.

File: hpx/parallel/executors/distribution_policy_executor.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/parallel/executors/distribution_policy_executor.hpp]: As long as we support launching both, Actions and non-Actions this should stay as it is.

File: hpx/parallel/executors/distribution_policy_executor.hpp (hunk 1 above)
Comment (sithhell) [inline hpx/parallel/executors/distribution_policy_executor.hpp]: Yeah, it just occurred to me when reading through the code

File: hpx/parallel/executors/distribution_policy_executor.hpp (hunk 2 above)
Comment (sithhell) [inline hpx/parallel/executors/distribution_policy_executor.hpp]: That's what I'm questioning here (though unrelated to that PR): does it make sense to call non-actions with distribution policies?
